### ISSUE TITLE
Inconsistent diagnosis and how to improve MAPE

### ISSUE THREAD

**Issue by andrewng88** [facebook/prophet#1058]
Hi,

Thanks for this wonderful, I've experimented with LSTM and ARIMA and came back to this one as I got a hunch that it can be further optimized.

I have 2 years and 6 months of data. It's daily and I've resampled to weekly.

`df_cv = cross_validation(m, initial='547.85 days', period='91.3125 days', horizon = '182.625 days')`

Is this ok as I'm following this 'rule' :  
By default, the initial training period is set to three times the horizon, and cutoffs are made every half a horizon.

Accuracy is less than ~5% for the 1st month but it was inconsistent later on as compared to Manning's data.

https://i.imgur.com/j6ox8tH.png
Is underfit/overfitting observed?

https://i.imgur.com/kME5xat.png

How do I increase the sensitivity to those spikes that I'm having. As in is possible to reduce the MAPE by another 1%, Settings are all default at the moment?

Thanks,

**Comment by bletham** [comment]
In this case the forecast looks pretty decent. If you look at the seasonal bump in July of the first year it predicts a little too low, but then in the 2nd July it predicts a little too high. Basically what the model is doing here is finding the seasonal effect that works best for both of them, which is something like the average of the two. That effect is consistent with multiplicative seasonality (the size of the seasonal effect decreases as the trend decreases), so I'd definitely try that and that may give some improvement: https://facebook.github.io/prophet/docs/multiplicative_seasonality.html

As for the cross validation, the reason it looks so wiggly is because it isn't using very many cutoffs to estimate prediction error, and so there is a lot of variance in those estimates. Initial is 1.5 years, and with a horizon of 0.5 years and total data length of 2.5 years, all of the cutoffs are between 1.5 years and 2.0 years, which with a period of 90 days means only 3 cutoffs (1.5 years, 1.75 years, 2.0 years). For comparison, the example in the documentation uses 11 cutoffs (it's a much longer time series).

What to do about that: The horizon is whatever you need to predict, and so if that's 6 months then leave it at 6 months. The initial should be long enough to capture the relevant dynamics of the time series, which here includes yearly seasonality so you could take that down to 1 year but no less. You can decrease `period` to get more forecasts, for instance using 30 days instead of 90 days. These error estimates will not be entirely independent, but I'd expect that you'll be able to reduce the variance somewhat by decreasing period to 30.

The final thing to note is that the blue line is just showing a rolling average of the gray points. The things suggested above will decrease the variance of the gray points, but if you want to additionally smooth out that estimate you can also increase the window size used for the rolling average as described at the bottom of https://facebook.github.io/prophet/docs/diagnostics.html, for instance to 0.2 would be a lot smoother while still giving some indiciation of how MAPE is changing with time.

**Comment by andrewng88** [comment]
Hi Ben, 

Thanks for your prompt reply.

I've adjusted to multiplicative based on your suggestion anda also the diagnostic to

`df_cv = cross_validation(m, initial='365.25 days', period='30 days', horizon = '90 days') `

Pls see the following images.

Multiplicative
https://i.imgur.com/36uouhq.png
https://imgur.com/1jhcdj0

Non-Multiplicative
https://i.imgur.com/dO0x0JZ.png
https://i.imgur.com/1jhcdj0.png

You're right, the Multiplicative is better by 0.05%.

Question
1) I notice that usually prophet gets really good out of the box with slight tweaking. usually is there a set of other parameters that one would tweak as I would still want to work on this.
2) may I know what is the coverage column part means?
3) for horizon, why does it starts after 18days. does it mean day 1 to day 18 has the same of better values?
4) side track. https://facebook.github.io/prophet/docs/trend_changepoints.html#automatic-changepoint-detection-in-prophet does changepoint means 'inflection point' like calculus. Meaning we need to increase it to have more flexibility( faster direction change? ).

Thank you.

**Comment by bletham** [comment]
1. The main parameters that are valuable to tune are "changepoint_prior_scale" and "seasonality_prior_scale". Next to those, I think the most value comes from looking at the forecast and seeing if there are holidays or similar consistent effects that are being missed and that could be added to improve things.

2. "Coverage" is percent of times that the point fell within the uncertainty interval [yhat_lower, yhat_upper]. By default it will be an 80% interval, and so if it is well-calibrated you should find that the coverage should be around 0.8.

3. This is because all of these metrics are computed as averages over a bunch of points. MAPE means average the absolute % error over a bunch of points. The number of points is determined by the `rolling_window` argument, which by default is set to 10%. So, in that dataframe the value of MAPE for h=18 is actually the mean APE across all of the predictions made with horizon 18 or less. It is a rolling window, so similarly the MAPE value at h=32 will be computed as the average APE over the previous 18 days.

4. In the calculus sense an inflection point means the 2nd derivative changes sign; here a changepoint just means that the slope changed value. In the case of the default piecewise linear trend, the first derivative is a piecwise constant function and it would be a step-change in the 1st derivative. But yes, increasing the changepoint flexibility (by increasing the changepoint_prior_scale and/or the number of changepoint points) will allow the trend to change more rapidly.

**Comment by andrewng88** [comment]
Sorry for the late reply.

1. For example, I checked with a domain expert and the advise is to use data after say 2015, the MAPE improved by x%. So using the correct data for modelling can be a hyperparameter too? 
What about say patterns that are not unidentifiable, ( i.e year T is trending up, year T-1 is trending down is there anything that can be done?

2. If coverage 1 means all the points are within the upper and lower boundary. If the coverage under 0.5 would the prediction still be useable?

3. Sorry I'm trying to understand properly.
https://facebook.github.io/prophet/docs/diagnostics.html
For this example, h = 365, 10%  of h is rolling window = 36, hence the first value is 37 days. _On this 8 year time series, this corresponds to 11 total forecasts._ Does it mean the MAPE value of 37 days are the mean of the 11 forecasts?

If the above understanding correct, if my requirement is monthly forecast the MAPE would be ~ 6% and if it's 1 year it will increase to ~ 10%

Question : 
what if my MAPE from 37 to 365 days are see saw ( very inconsistent/not stable at all )?does it mean it's overfitting/outlier that may affect? It's some sort like we do cross val score with sklearn, suddenly there is spike.

Thanks.

**Comment by bletham** [comment]
1. If there are changes in trend from year to year that are not explainable in some way, basically what will happen is you will have wider uncertainty bands for the trend in the future. Basically the model will assume that if the trend has changed a bunch in the past, then it might change a bunch in the future too and will bake that into the uncertainty bands (https://facebook.github.io/prophet/docs/uncertainty_intervals.html#uncertainty-in-the-trend)

2. Coverage is really about how much you should trust the uncertainty intervals. By default the uncertainty intervals are 80% intervals, meaning that if the model were perfectly calibrated, 80% of the time the data would fall within the interval. So if we measure the coverage as 1, that's bad because it means the uncertainty interval is too wide; we asked for 80% and we got 100%. If it is 0.5, it is similarly too small. How big of a deal is that? Well it depends. It means that when looking at the uncertainty intervals for the future, instead of believing that it should fall within the uncertainty 80% of the time, you know that it might be more like 50%. But that could still be OK; if the uncertainty intervals are too narrow but the main prediction has low error (e.g. MAPE), it could definitely still be useful.

3. This is a bit confusing. What is happening there is that MAPE is averaged over 11 forecasts for each of the h values, and on top of that is additionally averaged over the 37 days. So there's 37*11 values being included in each average. That's why the line is so smooth. If you're seeing it not smooth, then yes outliers / one day that is really hard to predict would be a likely explanation. You should be able to see this if you look at the gray dots in the plot, which show that not-averaged values (so for MAPE, it shows the % error at every point in `df_cv`. If there are a bunch of really high values in there, then those could be outliers that would then mess up the average.

**Comment by bletham** [comment]
3. continued: You could also then look in `df_cv` to see what days you are getting high errors for, and see if it is because of bad data on those days.

**Comment by andrewng88** [comment]
Dear @bletham ,

Thanks for the reply. 

Sorry for so many questions but I really want to understand more about this library.

1) How do you actually calibrate such that the coverage is around 80%. As per the manning dataset, we are only getting approximately 60% but it's consistent unlike mine.

I have 3 1/2 years worth of monthly data and I'm getting the following. 

Train on 1095 days
period 60 days
horizon 120days 
https://i.imgur.com/kov0x4Z.png

I tweak a little bit and I obtain the following.

https://i.imgur.com/IzoO3eF.png

I also observed that the mape is not as consistent as the manning dataset. I guess this is the inherent noise in the data that I'm working on.

how do I get a more consistent coverage?

2) If say, for this particular 1 month, sales are increased due to Olympics. How do you approach this since it's a 4 years affair, do you use the external regressor, is_olympics = 1 and 0 for the rest?

3) https://i.imgur.com/kTrvPZZ.png
Please refer to this image, if say some SOP has been changed in Dec 2018, would external regressor be useful or should I increase changepoint to 0.1? Would increasing changepoint be overfit?

4) Will it be useful to include holiday if my data is monthly/weekly or it must be daily/hourly to gain the holiday effect?

Thanks in advance.

**Comment by GaglianoM** [comment]
Hi @andrewng88,

I too am currently working on tuning my diagnostics (specifically, MAPE) for a model so this issue is quite relevant to me for similar reasons of inconsistency.

In response to your recent post to achieve more consistent coverage and looking at your plot, I would suggest definitely adding in holidays (country specific and custom dates) to help control for some of the trend/seasonality. 

Modifying your changepoint_prior_scale from the default 0.5 value to something larger such as 0.1 or 0.5 will increase the flexibility of the model. If you know of specific changepoints that aren't being properly recognized, I would suggest adding them custom as well. In fact, mine is currently set at 0.8! It may go down as I continue to adjust parameters.

Once you've iterated on these things I would (hopefully) expect a more consistent model - but it will certainly depend on proper application of the domain expertise as well. 

Best,

Michael

**Comment by andrewng88** [comment]
@GaglianoM 

Thanks for sharing. 

I'm currently using 0.1 for changepoint_prior_scale , I'm not sure if using 0.8 is ok as it might overfit but best to check with validation/diagnostics. 

For example if changing the parameter improves the MAPE slightly for me I'd leave it as defauly values.

So far what I did and would like to have @bletham chime in when he is free, I'd investigate the df.plot and see if there is any significant trend difference and combine it with domain expertise I will determine if I were to train less older data which is not relevant anymore. 

i.e after 2010, operation managed to do X activity to reduce expenditure by $100k .

**Comment by bletham** [comment]
Here are some comments on your earlier post:

1. There are a few reasons why the model is likely misspecified and so coverage won't equal 0.8 in general. One is the future trend uncertainty; the model assumes the future will see the same rate/magnitude of trend changes as the past. This assumption is probably not completely accurate, and so coverage will not be exactly the requested value. Another piece of model misspecification is that it assumes after accounting for trend/seasonality/regressors, all leftover variance is normal i.i.d. noise. This assumption is probably generally not exactly correct either: for instance in your plots you can see some trends in the "noise" where there are sequences of points that are consistently below or above the estimate that are inconsistent with the residuals being i.i.d noise. This will also impact the coverage. I think the right way to think of this is to take your empirically measured coverage (with cross validation) as the coverage you should expect moving forward. This is more likely to be true than the modeling assumptions. But like I said above, poor coverage doesn't mean the model predictions won't be good enough to be useful.

2. Yes, using a binary extra regressor for that sort of thing makes a lot of sense to me.

3. If you increase the changepoint prior scale, it will eventually give the model enough flexibility to fit that pattern. That might work here; as you note, increasing the changepoint prior scale does add some risk of overfitting because maybe you know this is a one-time thing and you don't _want_ the model to have that much flexiblity in the trend in the future. But in this case I'd guess that you could get decent results by increasing it enough to capture that trend change, and you could use cross validation (starting after Dec 2018) to make sure predictions moving forward are still reasonable.

4. If you specify a holiday effect it will have to match up with that same day in your data; meaning if your monthly data are aggregated to the last day of the month (like Dec 31 for December) and you want to include the holiday effect of Christmas, then you would need to put it on December 31 (the day in the history that includes data for Christmas), and not on December 25 (which is not in the history). Since this will anyway require manually specifying a bunch of dates, you might find it easier to just construct a binary extra regressors like in (2) and do it that way.
